### PR TITLE
feat(dataset): parallel data download/upload

### LINKED ISSUE #3282: Parallel push/pull of data
Currently, we download/upload a single file at a time when pulling/pushing data from a dataset with cloud storage backend. We should support multiple downloads/uploads in parallel.

### PR DESCRIPTION
**Pull request by mohammad-alisafaee**
# Description

Parallel pull for datasets with cloud storage and parallel add for all datasets. Moves some utility code around.

Fixes #3282

### CODE REVIEW

File: renku/core/dataset/providers/web.py [hunk 1]
```diff
@@ -161,37 +157,15 @@ def download_files(
...
-                delay=max_workers,  # NOTE: Rate limit to 1 request/second
-            )
-            for url, name in zip(urls, names)
-        }
-
-        for future in concurrent.futures.as_completed(futures):
-            files.extend(future.result())
-
-    return files
+    def download_file_wrapper(uri_filename: Tuple[str, Optional[str]], **kwargs):
```
Comment (Panaetius) [inline renku/core/dataset/providers/web.py:160]: if you changed it do `*d` in `parallel_execute` this wrapper wouldn't be necessary, I think

File: renku/domain_model/constant.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright 2019-2023 - Swiss Data Science Center (SDSC)
+# Copyright 2017-2023 - Swiss Data Science Center (SDSC)
```
Comment (Panaetius) [inline renku/domain_model/constant.py:2]: We should get rid of this/use the new header

File: renku/core/util/util.py [hunk 3]
```diff
@@ -34,8 +38,63 @@ def to_semantic_version(value: str) -> Optional[Version]:
...
+    """
+    try:
+        uuid_obj = uuid.UUID(value, version=4)
+    except ValueError:
+        return False
+
+    return str(uuid_obj) == value
+
+
+def parallel_execute(function: Callable[..., List[Any]], *, data: List[Any], rate: float = 1, **kwargs) -> List[Any]:
```
Comment (Panaetius) [inline renku/core/util/util.py:54]: Should we have a way to un-parallelize this in tests? Does it cause error to be swallowed/more difficult to debug?

**Review by Panaetius — CHANGES_REQUESTED** [review]
Really nice changes, having a default implementation for parallel execution is great!

File: renku/core/dataset/dataset.py [hunk 4]
```diff
@@ -1250,74 +1253,89 @@ def should_include(filepath: Path) -> bool:
...
-        path.parent.mkdir(parents=True, exist_ok=True)
-        # NOTE: Don't check if destination exists. ``IStorage.copy`` won't copy a file if it exists and is not modified.
+    return []
 
-        if not file.based_on:
-            raise errors.DatasetImportError(f"Dataset file doesn't have a URI: {file.entity.path}")
 
-        with communication.busy(f"Copying {file.entity.path} ..."):
-            storage.download(file.based_on.url, path)
+def pull_cloud_storage(name: str, location: Optional[Path] = None) -> None:
```
[the remark this answers is not in the document]
Comment (mohammad-alisafaee) [reply renku/core/dataset/dataset.py:1309]: The function had it but it got misplaced.

File: renku/core/dataset/dataset.py [hunk 5]
```diff
@@ -1250,74 +1253,89 @@ def should_include(filepath: Path) -> bool:
...
+    finally:
+        communication.finalize_progress(name=file.entity.path)
+
+    # NOTE: Update information that weren't present when adding file to the dataset
+    if not file.has_valid_checksum() or not file.has_valid_size():
+        if not file.has_valid_checksum():
+            md5_hash = hash_file(path, hash_type="md5") or ""
+            entity = Entity(path=file.entity.path, checksum=md5_hash)
+            remote_entity = RemoteEntity(checksum=md5_hash, url=file.based_on.url, path=file.based_on.path)
         else:
```
[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/core/dataset/dataset.py:1289]: but right now it's
```
if not a or not b:
    if not a:
        [...]
    else:
        [...]
```

and I just pointed out that that's the same as
```
if not a:
    [...]
elif not b:
    [...]
```
but with less nesting

File: renku/core/util/util.py [hunk 6]
```diff
@@ -34,8 +38,63 @@ def to_semantic_version(value: str) -> Optional[Version]:
...
+
+    Returns:
+        List(Any): A list of return results of all executions.
+
+    """
+    from renku.core.util import communication
+    from renku.core.util.contexts import wait_for
+    from renku.domain_model.project_context import project_context
+
+    assert rate > 0, f"Rate must be positive: {rate}"
```
[the remark this answers is not in the document]
Comment (mohammad-alisafaee) [reply renku/core/util/util.py:70]: We should merge that PR first then.

File: renku/core/dataset/dataset.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mohammad-alisafaee) [reply renku/core/dataset/dataset.py]: We can write in two separate ifs but then we need to duplicate `DatasetFile` instantiation code.

[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/core/dataset/dataset.py]: The way it is now is fine

**Review by Panaetius — DISMISSED** [review]

**Review by Panaetius — APPROVED** [review]

**Pull request merged**